cli/azd/pkg/project/ai_helper.go

```diff
@@ -0,0 +1,713 @@
+		deploymentName = fmt.Sprintf("%s-deployment", serviceConfig.Name)
+	}
+
+	deploymentName = fmt.Sprintf("%s-%d", deploymentName, a.clock.Now().Unix())
+
+	yamlFilePath := filepath.Join(serviceConfig.Path(), config.Deployment.Path)
+	_, err = os.Stat(yamlFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	a.env.DotenvSet(AiEndpointEnvVarName, endpointName)
+	a.env.DotenvSet(AiDeploymentEnvVarName, deploymentName)
+
+	deploymentArgs := []string{
+		"-t", "online-deployment",
+		"-s", scope.SubscriptionId(),
+		"-g", scope.ResourceGroup(),
+		"-w", scope.Workspace(),
+		"-f", yamlFilePath,
+		"--set", fmt.Sprintf("name=%s", deploymentName),
+		"--set", fmt.Sprintf("endpoint_name=%s", endpointName),
+	}
+
+	if environmentVersionName != "" {
+		deploymentArgs = append(deploymentArgs, "--set", fmt.Sprintf("environment=%s", environmentVersionName))
+	}
+
+	if modelVersionName != "" {
+		deploymentArgs = append(deploymentArgs, "--set", fmt.Sprintf("model=%s", modelVersionName))
+	}
+
+	deploymentArgs, err = a.applyOverrides(deploymentArgs, config.Deployment.Overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = a.pythonBridge.Run(ctx, ai.MLClient, deploymentArgs...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Poll for deployment and wait till it reaches a terminal state
+	onlineDeployment, err := a.pollForDeployment(ctx, scope, endpointName, deploymentName)
+	if err != nil {
+		return nil, err
+	}
+
+	return onlineDeployment, nil
+}
+
+// DeletePreviousDeployments deletes all previous deployments of an online endpoint except the one with 100% traffic
+func (a *aiHelper) DeletePreviousDeployments(
+	ctx context.Context,
+	scope *ai.Scope,
+	endpointName string,
+) error {
+	// Get the endpoint
+	getEndpointResponse, err := a.endpointsClient.Get(ctx, scope.ResourceGroup(), scope.Workspace(), endpointName, nil)
+	if err != nil {
+		return err
+	}
+
+	onlineEndpoint := getEndpointResponse.OnlineEndpoint
+	var deploymentName string
+
+	// Detect the deployment with 100% traffic
+	for key, trafficWeight := range onlineEndpoint.Properties.Traffic {
+		if *trafficWeight == 100 {
+			deploymentName = key
+			break
+		}
+	}
+
+	if deploymentName == "" {
+		return errors.New("no deployment found with 100% traffic")
+	}
+
+	// Get existing deployments
+	existingDeployments := []*armmachinelearning.OnlineDeployment{}
+
+	deploymentsPager := a.deploymentsClient.NewListPager(scope.ResourceGroup(), scope.Workspace(), endpointName, nil)
+	for deploymentsPager.More() {
+		page, err := deploymentsPager.NextPage(ctx)
+		if err != nil {
+			return err
+		}
+
+		existingDeployments = append(existingDeployments, page.Value...)
+	}
+
+	// Delete previous deployments
+	for _, existingDeployment := range existingDeployments {
+		// Ignore the new deployment
+		if *existingDeployment.Name == deploymentName {
+			continue
+		}
+
+		// We will start the delete operation but not wait for completion
+		_, err := a.deploymentsClient.BeginDelete(
```

Comment:
Just wondering - should we be adding these pollers to a group and then polling all of them at the end, so we can observe failures and report them and so callers of `DeletePreviousDeployments` know that the unused deployments are now gone instead of just in the process of being deleted?

Reply:
We could - just didn't want to make the deployment through `azd` to appear to take longer than we really need.

Reply:
Each deployment can take minutes to complete so I'd rather not take this time hit.  Keeping as is for now.